src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -73,16 +77,27 @@ internal sealed partial class LocalRewriter : BoundTreeRewriterWithStackGuard
             try
             {
                 var factory = new SyntheticBoundNodeFactory(method, statement.Syntax, compilationState, diagnostics);
+                DynamicAnalysisInjector dynamicInstrumenter = instrumentForDynamicAnalysis ? new DynamicAnalysisInjector(method, statement, factory, compilationState, diagnostics, debugDocumentProvider, Instrumenter.NoOp) : null;
 
                 // We don’t want IL to differ based upon whether we write the PDB to a file/stream or not.
                 // Presence of sequence points in the tree affects final IL, therefore, we always generate them.
                 var localRewriter = new LocalRewriter(compilation, method, methodOrdinal, containingType, factory, previousSubmissionFields, allowOmissionOfConditionalCalls, diagnostics,
-                                                      DebugInfoInjector.Singleton);
+                                                      instrumentForDynamicAnalysis ? new DebugInfoInjector(dynamicInstrumenter) : DebugInfoInjector.Singleton);
 
                 var loweredStatement = (BoundStatement)localRewriter.Visit(statement);
                 sawLambdas = localRewriter._sawLambdas;
                 sawLocalFunctions = localRewriter._sawLocalFunctions;
                 sawAwaitInExceptionHandler = localRewriter._sawAwaitInExceptionHandler;
+
+                if (instrumentForDynamicAnalysis)
+                {
+                    BoundBlock loweredBlock = loweredStatement as BoundBlock;
+                    if (loweredBlock != null)
+                    {
+                        loweredStatement = dynamicInstrumenter.AddInstrumentationPrologue(loweredBlock, ref dynamicAnalysisSpans);
```

Comment:
After some discussion and adjustment, CreateBlockPrologue appears to meet the requirements.